narwhal/executor/src/subscriber.rs

```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
-                return Some(batch);
+        let request_batches_response = self.network.request_batches(digests, worker).await;
+        let mut response_batches: HashMap<BatchDigest, Batch> = HashMap::new();
+
+        match request_batches_response {
+            Ok(fetched_batches) => {
+                debug!(
+                    "Found {} batches locally: {:?}",
+                    fetched_batches.len(),
+                    fetched_batches
+                );
+                for fetched_batch in fetched_batches {
+                    self.metrics.successful_local_batch_fetch.inc();
+                    response_batches.insert(fetched_batch.digest(), fetched_batch);
+                }
+            }
+            Err(err) => {
+                if err.to_string().contains("Timeout") {
+                    self.metrics.local_batch_fetch_timeout.inc();
+                } else {
+                    self.metrics.failed_local_batch_fetch.inc();
+                }
+                warn!("Error communicating with our own worker: {err}");
             }
-            Ok(None) => debug!("Payload {} not found locally", digest),
-            Err(err) => warn!("Error communicating with own worker: {}", err),
         }
-        None
+
+        response_batches
     }
 
-    /// This future performs fetch from given worker
+    /// This future performs a fetch from a given remote worker
     /// This future performs infinite retries with exponential backoff
     /// You can specify stagger_delay before request is issued
-    #[instrument(level = "debug", skip_all, fields(stagger_delay = ? stagger_delay, worker = % worker, digest = % digest))]
-    async fn fetch_from_worker(
+    #[instrument(level = "debug", skip_all, fields(stagger_delay = ? stagger_delay, worker = % worker, digests = ? digests))]
+    async fn fetch_remote(
         &self,
         stagger_delay: Duration,
         worker: NetworkPublicKey,
-        digest: BatchDigest,
-    ) -> Batch {
+        digests: HashSet<BatchDigest>,
+    ) -> HashMap<BatchDigest, Batch> {
         tokio::time::sleep(stagger_delay).await;
         // TODO: Make these config parameters
         let max_timeout = Duration::from_secs(60);
         let mut timeout = Duration::from_secs(10);
         let mut attempt = 0usize;
+        let mut remaining_digests = digests.clone();
+        let mut batches: HashMap<BatchDigest, Batch> = HashMap::new();
         loop {
             attempt += 1;
             let deadline = Instant::now() + timeout;
             let request_batch_guard =
                 PendingGuard::make_inc(&self.metrics.pending_remote_request_batch);
-            let payload =
-                tokio::time::timeout_at(deadline, self.safe_request_batch(digest, worker.clone()))
-                    .await;
+            let response = tokio::time::timeout_at(
+                deadline,
+                self.safe_request_batches(remaining_digests.clone(), worker.clone()),
```

Comment:
use anemo with_timeout on the rpc, rather than wrapping in tokio timeout?

this will require you to change your code in p2p.rs slightly to have the trait method take `request: impl anemo::types::request::IntoRequest<FetchBatchesRequest> + Send,`

example usage of with_timeout in narhwal:
https://github.com/MystenLabs/sui/blob/aad6425854d725a28ffee7d819b35913346ef212/narwhal/primary/src/synchronizer.rs#L773